crates/bevy_ecs/macros/src/component.rs

```diff
@@ -58,11 +58,34 @@ pub fn derive_component(input: TokenStream) -> TokenStream {
     let struct_name = &ast.ident;
     let (impl_generics, type_generics, where_clause) = &ast.generics.split_for_impl();
 
-    TokenStream::from(quote! {
-        impl #impl_generics #bevy_ecs_path::component::Component for #struct_name #type_generics #where_clause {
-            const STORAGE_TYPE: #bevy_ecs_path::component::StorageType = #storage;
-        }
-    })
+    #[cfg(not(feature = "bevy_reflect"))]
+    {
+        TokenStream::from(quote! {
+            impl #impl_generics #bevy_ecs_path::component::Component for #struct_name #type_generics #where_clause {
+                const STORAGE_TYPE: #bevy_ecs_path::component::StorageType = #storage;
+            }
+        })
+    }
+    #[cfg(feature = "bevy_reflect")]
+    {
+        let register_type = if has_reflect_attr(&ast) {
+            quote! {
+                #[doc(hidden)]
+                fn __register_type(registry: &#bevy_ecs_path::private::bevy_reflect::TypeRegistryArc) {
+                    registry.write().register::<Self>();
+                }
+            }
+        } else {
+            proc_macro2::TokenStream::new()
+        };
```

Comment:
Nit: I think this can be further simplified:

```rust
let mut register_type: Option<proc_macro2::TokenStream> = None;

#[cfg(feature = "bevy_reflect")]
if has_reflect_attr(&ast) {
  register_type = Some(...);
}

TokenStream::from(quote! {
  impl #impl_generics #bevy_ecs_path::component::Component for #struct_name #type_generics #where_clause {
    const STORAGE_TYPE: #bevy_ecs_path::component::StorageType = #storage;

    #register_type
  }
})
```

Reply:
Done!.